src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,217 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.Objects;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+
+    class Node<E> {
+        private E item;
+        private Node<E> next;
+        private Node<E> prev;
+
+        public Node(Node<E> prev, E element, Node<E> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    public MyLinkedList() {
+    }
+
     @Override
     public void add(T value) {
+        if (head == null) {
+            head = new Node<>(null, value, null);
+            tail = head;
+        } else {
+            Node<T> newNode = new Node<>(tail, value, null);
+            tail.next = newNode;
+            tail = newNode;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size()) {
+            throw new IndexOutOfBoundsException("Index out of bounds: " + index);
+        }
+        if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, head);
+            head = newNode;
+            if (tail == null) {
+                tail = newNode;
+            }
+        } else {
+            Node<T> current = head;
+            int count = 1;
+            while (current != null) {
+                if (count == index) {
+                    Node<T> newNode = new Node<>(current, value, current.next);
+                    current.next = newNode;
+                    current.prev = newNode;
+                    if (newNode.next == null) {
+                        tail = newNode;
+                    }
+                    break;
+                } else {
+                    count++;
+                    current = current.next;
+                }
+            }
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list.isEmpty()) {
+            return;
+        }
+
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
```

Comment:
create getNodeByIndex method, move checkIndex there, iterate from closest end of list (head or tail)